### PR TITLE
[review] Allow a client to send "remote" commands programatically

### PR DESCRIPTION
**Pull request by krocard**
Users can already send commands to a Parameter Framework instance through a
socket (provided that they asked for this interface to be started). Some
features are only available through this interface.

In upcoming patches, we want to use those features and send such commands but
we don't want to use a remote interface, which is sometimes painful. This patch
makes the "remote" commands available through a programmatic API. This API
expects a command name and a string-vector containing the arguments; which
makes it fairly easy to use and to implement a user interface providing the
same commands as they are used to - but without the need of a socket.

The object that provides this API is allocated by the Parameter Framework but
is owned by the client - we want to be C++03-retrocompatible, which forbids us
from returning a unique_ptr, this is why we return a raw pointer.
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/01org/parameter-framework/pull/324%23issuecomment-167792834%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/324%23issuecomment-167803236%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/324%23issuecomment-167835831%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/324%23issuecomment-167965535%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/324%23issuecomment-167989858%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/324%23issuecomment-167992135%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/324%23issuecomment-168004530%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/324%23discussion_r48606231%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/324%23discussion_r48606236%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/324%23discussion_r48606268%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/324%23discussion_r48606274%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/324%23discussion_r48607804%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/324%23discussion_r48607806%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/324%23discussion_r48607807%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/324%23discussion_r48607808%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/01org/parameter-framework/pull/324%23issuecomment-167792834%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Taking%20over%20%23321%20from%20%40dawagner%20%22%2C%20%22created_at%22%3A%20%222015-12-29T13%3A40%3A51Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3Ashipit%3A%20%22%2C%20%22created_at%22%3A%20%222015-12-29T14%3A48%3A10Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/11178583%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tcahuzax%22%7D%7D%2C%20%7B%22body%22%3A%20%22Need%20to%20fix%20https%3A//github.com/dawagner/parameter-framework/commit/b2a868693625b82d2e4b02ba1667e9a20267ff10%22%2C%20%22created_at%22%3A%20%222015-12-29T17%3A32%3A07Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%2C%20%7B%22body%22%3A%20%22format%20with%20clang%5Cr%5Cn%3A-1%3A%20%22%2C%20%22created_at%22%3A%20%222015-12-30T09%3A28%3A54Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/11178583%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tcahuzax%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%20Formated%20with%20clang-format%22%2C%20%22created_at%22%3A%20%222015-12-30T12%3A30%3A27Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%20%3Ashipit%3A%20%22%2C%20%22created_at%22%3A%20%222015-12-30T12%3A56%3A31Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3Ashipit%3A%20%22%2C%20%22created_at%22%3A%20%222015-12-30T14%3A03%3A27Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/11178583%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tcahuzax%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20f6cf674c70145c417362601b4c6ac19b18826c82%20tools/xmlGenerator/domainGeneratorConnector.cpp%20231%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/01org/parameter-framework/pull/324%23discussion_r48606268%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22catch%20any%20exception%22%2C%20%22created_at%22%3A%20%222015-12-30T14%3A29%3A16Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-none%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222015-12-30T14%3A56%3A38Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20tools/xmlGenerator/domainGeneratorConnector.cpp%3AL1-236%22%7D%2C%20%22Pull%20f6cf674c70145c417362601b4c6ac19b18826c82%20tools/xmlGenerator/domainGeneratorConnector.cpp%20230%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/01org/parameter-framework/pull/324%23discussion_r48606231%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22return%20normalizeExitCode%28errorNb%29%22%2C%20%22created_at%22%3A%20%222015-12-30T14%3A28%3A37Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-none%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222015-12-30T14%3A56%3A39Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20tools/xmlGenerator/domainGeneratorConnector.cpp%3AL1-236%22%7D%2C%20%22Pull%20f6cf674c70145c417362601b4c6ac19b18826c82%20tools/xmlGenerator/domainGeneratorConnector.cpp%20233%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/01org/parameter-framework/pull/324%23discussion_r48606236%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22return%20normalizeExitCode%28errorNb%20%2B%201%29%22%2C%20%22created_at%22%3A%20%222015-12-30T14%3A28%3A43Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-none%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222015-12-30T14%3A56%3A39Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20tools/xmlGenerator/domainGeneratorConnector.cpp%3AL1-236%22%7D%2C%20%22Pull%20f6cf674c70145c417362601b4c6ac19b18826c82%20tools/xmlGenerator/domainGeneratorConnector.cpp%20197%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/01org/parameter-framework/pull/324%23discussion_r48606274%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22static%20inline%22%2C%20%22created_at%22%3A%20%222015-12-30T14%3A29%3A25Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-complete%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222015-12-30T14%3A56%3A39Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20tools/xmlGenerator/domainGeneratorConnector.cpp%3AL1-236%22%7D%7D%2C%20%22approved%22%3A%20%7B%22https%3A//github.com/krocard%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%7D%2C%20%22https%3A//github.com/tcahuzax%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/11178583%3Fv%3D3%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>

<a href='https://github.com/krocard'></a><a href='https://github.com/tcahuzax'></a>

- [x] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/01org/parameter-framework/pull/324#issuecomment-167792834'>General Comment</a></b>
- <a href='https://github.com/krocard'></a> Taking over #321 from @dawagner
- <a href='https://github.com/tcahuzax'></a> :shipit:
- <a href='https://github.com/krocard'></a> Need to fix https://github.com/dawagner/parameter-framework/commit/b2a868693625b82d2e4b02ba1667e9a20267ff10
- <a href='https://github.com/tcahuzax'></a> format with clang
:-1:
- <a href='https://github.com/krocard'></a> :+1: Formated with clang-format
- <a href='https://github.com/krocard'></a> :+1: :shipit:
- [x] <a href='#crh-comment-Pull f6cf674c70145c417362601b4c6ac19b18826c82 tools/xmlGenerator/domainGeneratorConnector.cpp 230'></a> &nbsp;<b><a href='https://github.com/01org/parameter-framework/pull/324#discussion_r48606231'>File: tools/xmlGenerator/domainGeneratorConnector.cpp:L1-236</a></b>
- <a href='https://github.com/krocard'></a> return normalizeExitCode(errorNb)
- [x] <a href='#crh-comment-Pull f6cf674c70145c417362601b4c6ac19b18826c82 tools/xmlGenerator/domainGeneratorConnector.cpp 233'></a> &nbsp;<b><a href='https://github.com/01org/parameter-framework/pull/324#discussion_r48606236'>File: tools/xmlGenerator/domainGeneratorConnector.cpp:L1-236</a></b>
- <a href='https://github.com/krocard'></a> return normalizeExitCode(errorNb + 1)
- [x] <a href='#crh-comment-Pull f6cf674c70145c417362601b4c6ac19b18826c82 tools/xmlGenerator/domainGeneratorConnector.cpp 231'></a> &nbsp;<b><a href='https://github.com/01org/parameter-framework/pull/324#discussion_r48606268'>File: tools/xmlGenerator/domainGeneratorConnector.cpp:L1-236</a></b>
- <a href='https://github.com/krocard'></a> catch any exception
- [x] <a href='#crh-comment-Pull f6cf674c70145c417362601b4c6ac19b18826c82 tools/xmlGenerator/domainGeneratorConnector.cpp 197'></a> &nbsp;<b><a href='https://github.com/01org/parameter-framework/pull/324#discussion_r48606274'>File: tools/xmlGenerator/domainGeneratorConnector.cpp:L1-236</a></b>
- <a href='https://github.com/krocard'></a> static inline

<a href='https://www.codereviewhub.com/01org/parameter-framework/pull/324?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/01org/parameter-framework/pull/324?approve=1'></a>&nbsp;<a href='https://www.codereviewhub.com/01org/parameter-framework/pull/324?approve=0'></a>&nbsp;<a href='https://github.com/01org/parameter-framework/pull/324'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by dawagner** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/01org/parameter-framework/pull/324'>Refresh</a>.<a href='#crh-update-none'></a>

**Comment by krocard** [comment]
Taking over #321 from @dawagner

**Comment by tcahuzax** [comment]
:shipit:

**Comment by krocard** [comment]
Need to fix https://github.com/dawagner/parameter-framework/commit/b2a868693625b82d2e4b02ba1667e9a20267ff10

**Comment by tcahuzax** [comment]
format with clang
:-1:

**Comment by krocard** [comment]
:+1: Formated with clang-format

**Comment by krocard** [comment]
:+1: :shipit:

**Comment by tcahuzax** [comment]
:ship:

**Comment by tcahuzax** [comment]
:shipit:

File: tools/xmlGenerator/domainGeneratorConnector.cpp [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+                  << "    schemasDir=" << schemasDir << endl;
+    }
+
+    try {
+        XmlGenerator xmlGenerator(toplevelConfig, validate, verbose, schemasDir);
+        auto errorNb = xmlGenerator.parse(std::cin);
+        // TODO: add a check for conflicting elements
+        xmlGenerator.exportDomains(std::cout);
+
+        return int(std::min<decltype(errorNb)>(errorNb, std::numeric_limits<uint8_t>::max()));
```
Comment (krocard) [inline tools/xmlGenerator/domainGeneratorConnector.cpp]: return normalizeExitCode(errorNb)

File: tools/xmlGenerator/domainGeneratorConnector.cpp [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    } catch (runtime_error e) {
+        std::cerr << e.what() << std::endl;
+        return 1;
```
Comment (krocard) [inline tools/xmlGenerator/domainGeneratorConnector.cpp]: return normalizeExitCode(errorNb)

File: tools/xmlGenerator/domainGeneratorConnector.cpp [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+    }
+
+    try {
+        XmlGenerator xmlGenerator(toplevelConfig, validate, verbose, schemasDir);
+        auto errorNb = xmlGenerator.parse(std::cin);
+        // TODO: add a check for conflicting elements
+        xmlGenerator.exportDomains(std::cout);
+
+        return int(std::min<decltype(errorNb)>(errorNb, std::numeric_limits<uint8_t>::max()));
+    } catch (runtime_error e) {
```
Comment (krocard) [inline tools/xmlGenerator/domainGeneratorConnector.cpp]: catch any exception

File: tools/xmlGenerator/domainGeneratorConnector.cpp [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+Exit with the number of (recoverable or not error) that occured.
+
+This program is not intended to be used standalone but rather called through
+domainGenerator.py)";
+
+/** On linux at least, a program can not exit with a value greater than 255.
+ * @return min(code, 255);
+ */
+template <class T>
+int normalizeExitCode(T code)
```
Comment (krocard) [inline tools/xmlGenerator/domainGeneratorConnector.cpp]: static inline

File: tools/xmlGenerator/domainGeneratorConnector.cpp (hunk 3 above)
Comment (krocard) [inline tools/xmlGenerator/domainGeneratorConnector.cpp]: :+1:<a href='#crh-update-none'></a>

File: tools/xmlGenerator/domainGeneratorConnector.cpp (hunk 1 above)
Comment (krocard) [inline tools/xmlGenerator/domainGeneratorConnector.cpp]: :+1:<a href='#crh-update-none'></a>

File: tools/xmlGenerator/domainGeneratorConnector.cpp (hunk 2 above)
Comment (krocard) [inline tools/xmlGenerator/domainGeneratorConnector.cpp]: :+1:<a href='#crh-update-none'></a>

File: tools/xmlGenerator/domainGeneratorConnector.cpp (hunk 4 above)
Comment (krocard) [inline tools/xmlGenerator/domainGeneratorConnector.cpp]: :+1:<a href='#crh-update-complete'></a>

**Pull request merged**